circuitbreaker/circuit_breaker.go

```diff
@@ -0,0 +1,413 @@
+func (cb *CircuitBreaker) GetState() State {
+	cb.mutex.RLock()
+	defer cb.mutex.RUnlock()
+	return cb.state
+}
+
+// IsOpen returns true if the circuit is open
+func (cb *CircuitBreaker) IsOpen() bool {
+	return cb.GetState() == StateOpen
+}
+
+// Reset resets the circuit breaker to its initial closed state
+func (cb *CircuitBreaker) Reset() {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	// Reset counters
+	atomic.StoreInt64(&cb.failureCount, 0)
+	atomic.StoreInt64(&cb.successCount, 0)
+
+	// Reset state
+	cb.state = StateClosed
+	cb.lastStateChange = cb.now()
+
+	// Cancel any pending state transitions
+	if cb.openTimer != nil {
+		cb.openTimer.Stop()
+	}
+}
+
+// ForceOpen forcibly opens the circuit regardless of failure count
+func (cb *CircuitBreaker) ForceOpen() {
+	cb.transitionToOpen()
+}
+
+// ForceClose forcibly closes the circuit regardless of current state
+func (cb *CircuitBreaker) ForceClose() {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	cb.state = StateClosed
+	cb.lastStateChange = cb.now()
+	atomic.StoreInt64(&cb.failureCount, 0)
+	atomic.StoreInt64(&cb.successCount, 0)
+
+	if cb.openTimer != nil {
+		cb.openTimer.Stop()
+	}
+}
+
+// SetTimeout updates the timeout duration
+func (cb *CircuitBreaker) SetTimeout(timeout time.Duration) {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	cb.timeout = timeout
+}
+
+// transitionToOpen changes state to open and schedules transition to half-open
+func (cb *CircuitBreaker) transitionToOpen() {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	if cb.state != StateOpen {
+		cb.state = StateOpen
+		cb.lastStateChange = cb.now()
+
+		// Stop existing timer if any
+		if cb.openTimer != nil {
+			cb.openTimer.Stop()
+		}
+
+		// Schedule transition to half-open after timeout
+		cb.openTimer = time.AfterFunc(cb.timeout, func() {
+			cb.transitionToHalfOpen()
+		})
+
+		// Reset failure counter
+		atomic.StoreInt64(&cb.failureCount, 0)
+	}
+}
+
+// transitionToHalfOpen changes state from open to half-open
+func (cb *CircuitBreaker) transitionToHalfOpen() {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	if cb.state == StateOpen {
+		cb.state = StateHalfOpen
+		cb.lastStateChange = cb.now()
+
+		// Reset counters
+		atomic.StoreInt64(&cb.failureCount, 0)
+		atomic.StoreInt64(&cb.successCount, 0)
+
+		// Empty the semaphore channel
+		select {
+		case <-cb.halfOpenSemaphore:
+		default:
+		}
+	}
+}
+
+// transitionToClosed changes state from half-open to closed
+func (cb *CircuitBreaker) transitionToClosed() {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	if cb.state == StateHalfOpen {
+		cb.state = StateClosed
+		cb.lastStateChange = cb.now()
+
+		// Reset counters
+		atomic.StoreInt64(&cb.failureCount, 0)
+		atomic.StoreInt64(&cb.successCount, 0)
+	}
+}
+
+// AllowRequest determines if a request is allowed based on circuit state
+func (cb *CircuitBreaker) AllowRequest() (bool, State) {
+	atomic.AddInt64(&cb.totalRequests, 1)
+
+	cb.mutex.RLock()
```

Comment:
I think guarding this small block with read lock makes little difference. This whole method should be behind read lock if it is necessary that limits are honored/guarded against races during extreme peaks.